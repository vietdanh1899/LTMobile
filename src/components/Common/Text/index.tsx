import React, { PureComponent } from 'react';
import {
  Text as EText,
  Icon as EIcon,
  IconProps as EIconProps,
  TextProps as ETextProps,
  withTheme,
  ThemeProps,
} from 'react-native-elements';
import { StyleSheet, Platform } from 'react-native';
import { Fonts, lightTheme } from '@themes';
import QuickView from '@components/Common/View/QuickView';
import { Translation } from 'react-i18next';

type TFontFamily = typeof Fonts.fontFamily;
type TFontWeight = typeof Fonts.fontWeight;
type TFontSize = typeof Fonts.fontSize;
type TType = typeof lightTheme.Text;

export interface TextProps
  extends Omit<ETextProps, 'fontFamily' | 'fontWeight' | 'fontSize'> {
  fontFamily?: keyof TFontFamily;
  fontWeight?: keyof TFontWeight;
  fontSize?: keyof TFontSize | number;
  type?: keyof TType;
  primary?: boolean;
  secondary?: boolean;
  success?: boolean;
  warning?: boolean;
  error?: boolean;
  color?: string;
  center?: boolean;
  iconRight?: boolean;
  margin?: number;
  marginTop?: number;
  marginBottom?: number;
  marginLeft?: number;
  marginRight?: number;
  marginHorizontal?: number;
  marginVertical?: number;
  bold?: boolean;
  italic?: boolean;
  underline?: boolean;
  children?: any;
  icon?: EIconProps;
  iconContainerStyle?: any;
  t?: string;
  theme?: any;
}

class Text extends PureComponent<TextProps> {
  static defaultProps = {
    iconRight: false,
    type: 'paragraph',
  };

  render() {
    const {
      fontFamily,
      fontWeight,
      fontSize,
      primary,
      secondary,
      success,
      error,
      warning,
      color,
      center,
      margin,
      marginLeft,
      marginRight,
      marginBottom,
      marginTop,
      marginHorizontal,
      marginVertical,
      underline,
      bold,
      italic,
      icon,
      iconRight,
      iconContainerStyle,
      style,
      children,
      theme,
      type,
      t,
      ...otherProps
    } = this.props;

    /**
     * Color Handle
     */
    let colors: any;
    let { textColor } = theme.colors;
    if (Platform.OS === 'ios') {
      colors = theme.colors.platform.ios;
    } else {
      colors = theme.colors.platform.android;
    }
    if (primary) textColor = colors.primary;
    if (secondary) textColor = colors.secondary;
    if (success) textColor = colors.success;
    if (warning) textColor = colors.warning;
    if (error) textColor = colors.error;
    if (color) textColor = color;

    /**
     * textStyle
     */
    const textStyle = StyleSheet.flatten([
      type && theme.Text[type],
      fontFamily && {
        fontFamily: Fonts.fontFamily[fontFamily],
      },
      italic && {
        fontFamily: Fonts.fontFamily.RobotoItalic,
      },
      fontWeight && {
        fontWeight: Fonts.fontWeight[fontWeight],
      },
      bold && {
        fontWeight: Fonts.fontWeight.bold,
      },
      fontSize
        && typeof fontSize === 'number' && {
        fontSize,
      },
      fontSize
        && typeof fontSize === 'string' && {
        fontSize: Fonts.fontSize[fontSize],
      },
      center && {
        textAlign: 'center',
      },
      margin && { margin },
      marginTop && { marginTop },
      marginBottom && { marginBottom },
      marginLeft && { marginLeft },
      marginRight && { marginRight },
      marginHorizontal && { marginHorizontal },
      marginVertical && { marginVertical },
      underline && { textDecorationLine: 'underline' },
      { color: textColor },
      style,
    ]);

    /**
     * iconStyle
     */
    const iconStyle = StyleSheet.flatten([
      textStyle,
      { marginRight: iconRight ? 0 : 2 },
      { marginLeft: iconRight ? 2 : 0 },
      iconContainerStyle,
    ]);
    if (icon) {
      return (
        <QuickView
          testID="EIconText"
          row={!iconRight}
          rowReverse={iconRight}
          justifyContent={iconRight ? 'flex-end' : 'flex-start'}
        >
          <QuickView style={iconStyle}>
            <EIcon
              {...icon}
              name={icon.name}
              type={icon.type || 'material-community'}
              size={icon.size || 15}
              color={icon.color || textColor}
            />
          </QuickView>
          <QuickView>
            <EText {...otherProps} style={textStyle}>
              {children}
            </EText>
          </QuickView>
        </QuickView>
      );
    }

    if (t) {
      return (
        <Translation>
          {(trans) => (
            <EText {...otherProps} style={textStyle}>
              {trans(t)}
            </EText>
          )}
        </Translation>
      );
    }
    return (
      <EText {...otherProps} style={textStyle}>
        {children}
      </EText>
    );
  }
}

export default withTheme(
  Text as React.ComponentType<TextProps & ThemeProps<any>>,
);
