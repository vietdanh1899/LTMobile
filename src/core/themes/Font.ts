export const fontFamily = {
  RobotoBold: 'Roboto-Bold',
  RobotoItalic: 'Roboto-Italic',
  RobotoLight: 'Roboto-Light',
  RobotoRegular: 'Roboto-Regular',
  RobotoMedium: 'Roboto-Medium',
  RobotoThin: 'Roboto-Thin',
  GothamRoundedBold: 'GothamRounded-Bold',
  LatoBoldItalic: 'Lato-BoldItaclic',
  LatoBold: 'Lato-Bold',
};

export const fontWeight = {
  thin: '100',
  light: '300',
  medium: '600',
  bold: '700',
  heavy: '900',
};

export const fontSize = {
  tiny: 10,
  small: 12,
  medium: 14,
  large: 18,
  xLarge: 20,
};

export default { fontFamily, fontWeight, fontSize };
