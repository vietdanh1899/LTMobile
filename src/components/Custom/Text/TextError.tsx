import React, { PureComponent } from 'react';
import { TError } from '@utils/redux';
import _ from 'lodash';
import Text, { TextProps } from '../../Common/Text';
import QuickView from '../../Common/View/QuickView';

export interface TextErrorProps extends Omit<TextProps, 'error'> {
  error: null | TError;
}
class TextError extends PureComponent<TextErrorProps> {
  renderMessage = () => {
    const { error, ...otherProps } = this.props;
    if (error) {
      return error?.messages.map((message, index) => (
        <Text
          key={index.toString()}
          {...otherProps}
          marginVertical={2}
          center
          error
        >
          {message}
        </Text>
      ));
    }
    return null;
  };

  render() {
    const { error } = this.props;
    if (error && !_.isUndefined(_.head(error.messages))) {
      return <QuickView marginVertical={2}>{this.renderMessage()}</QuickView>;
    }
    return null;
  }
}

export default TextError;
