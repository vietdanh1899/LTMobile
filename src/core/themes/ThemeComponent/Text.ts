import { lightComponentColor, darkComponentColor } from './Common/CommonProps';

/**
 * Text
 */
const headerFontSize = 28;
const headerFontWeight = 'bold';
const xTitleFontSize = 24;
const titleFontSize = 20;
const paragraphFontSize = 16;
const subtitleFontSize = 14;

export const darkText = {
  header: {
    color: darkComponentColor.textColor,
    fontWeight: headerFontWeight,
    fontSize: headerFontSize,
  },
  xTitle: {
    color: darkComponentColor.textColor,
    fontSize: xTitleFontSize,
  },
  title: {
    color: darkComponentColor.textColor,
    fontSize: titleFontSize,
  },
  paragraph: {
    color: darkComponentColor.textColor,
    fontSize: paragraphFontSize,
  },
  subtitle: {
    color: darkComponentColor.textColorSecondary,
    fontSize: subtitleFontSize,
  },
};
export const lightText = {
  header: {
    color: lightComponentColor.textColor,
    fontWeight: headerFontWeight,
    fontSize: headerFontSize,
  },
  xTitle: {
    color: lightComponentColor.textColor,
    fontSize: xTitleFontSize,
  },
  title: {
    color: lightComponentColor.textColor,
    fontSize: titleFontSize,
  },
  paragraph: {
    color: lightComponentColor.textColor,
    fontSize: paragraphFontSize,
  },
  subtitle: {
    color: lightComponentColor.textColorSecondary,
    fontSize: subtitleFontSize,
  },
};
