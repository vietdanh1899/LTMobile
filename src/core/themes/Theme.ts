import {
  bottomNavigationBarHeight,
  headerPaddingHorizontal,
  bodyPaddingHorizontal,
  shadowViewLight,
  shadowViewDark,
  darkComponentColor,
  lightComponentColor,
} from './ThemeComponent/Common/CommonProps';
import {
  Color,
  darkPlatform,
  lightPlatform,
  darkPrimaryColor,
  darkSecondaryColor,
  lightPrimaryColor,
  lightSecondaryColor,
} from './ThemeComponent/Common/Color';
import { darkText, lightText } from './ThemeComponent/Text';
import { darkButton, lightButton } from './ThemeComponent/Button';
import { darkHeader, lightHeader } from './ThemeComponent/Header';
import { darkInput, lightInput } from './ThemeComponent/Input';
import { darkCheckBox, lightCheckBox } from './ThemeComponent/ListCheckBox';
import { darkDropdown, lightDropdown } from './ThemeComponent/Dropdown';
import { darkModal, lightModal } from './ThemeComponent/Modal';
import { lightParallaxScrollView, darkParallaxScrollView } from './ThemeComponent/ParallaxScrollView';

const ThemeEnum = {
  DARK: 'dark',
  LIGHT: 'light',
};

/**
 * Theme
 */
export const darkTheme = {
  key: ThemeEnum.DARK,
  colors: {
    primary: darkPrimaryColor,
    secondary: darkSecondaryColor,

    white: Color.white,
    grey1: Color.grey1,
    grey2: Color.grey2,
    grey3: Color.grey3,
    grey4: Color.grey4,
    grey5: Color.grey5,
    grey6: Color.grey6,
    grey7: Color.grey7,
    grey8: Color.grey8,
    grey10: Color.grey10,
    black: Color.black,

    greyOutline: '#bbb',
    searchBg: '#303337',

    success: Color.green,
    error: Color.red,
    warning: Color.yellow,
    disabled: 'hsl(208, 8%, 90%)',
    // Darker color if hairlineWidth is not thin enough
    divider: Color.grey6,
    border: Color.grey9,
    platform: darkPlatform,

    /**
    * Component Color
    */
    bgColor: darkComponentColor.bgColor,
    bgColorSecondary: darkComponentColor.bgColorSecondary,
    textColor: darkComponentColor.textColor,
    textColorSecondary: darkComponentColor.textColorSecondary,
  },
  /**
   * Component
   */
  bottomNavigationBarHeight,
  headerPaddingHorizontal,
  bodyPaddingHorizontal,
  shadowView: shadowViewDark,

  Text: darkText,
  Button: darkButton,
  Header: darkHeader,
  Input: darkInput,
  CheckBox: darkCheckBox,
  Dropdown: darkDropdown,
  AppModal: darkModal,
  ParallaxScrollView: darkParallaxScrollView,
};

export const lightTheme = {
  key: ThemeEnum.LIGHT,
  colors: {
    primary: lightPrimaryColor,
    secondary: lightSecondaryColor,

    white: Color.white,
    grey1: Color.grey1,
    grey2: Color.grey2,
    grey3: Color.grey3,
    grey4: Color.grey4,
    grey5: Color.grey5,
    grey6: Color.grey6,
    grey7: Color.grey7,
    grey8: Color.grey8,
    grey10: Color.grey10,
    black: Color.black,

    greyOutline: '#bbb',
    searchBg: '#303337',

    listItemBg: Color.white,
    success: Color.green,
    error: Color.red,
    warning: Color.yellow,
    disabled: 'hsl(208, 8%, 90%)',
    // Darker color if hairlineWidth is not thin enough
    divider: Color.grey2,
    border: Color.grey2,
    platform: lightPlatform,

    /**
     * Component Color
     */
    bgColor: lightComponentColor.bgColor,
    bgColorSecondary: lightComponentColor.bgColorSecondary,
    textColor: lightComponentColor.textColor,
    textColorSecondary: lightComponentColor.textColorSecondary,
  },

  /**
   * Component
   */
  bottomNavigationBarHeight,
  headerPaddingHorizontal,
  bodyPaddingHorizontal,
  shadowView: shadowViewLight,

  Text: lightText,
  Button: lightButton,
  Header: lightHeader,
  Input: lightInput,
  CheckBox: lightCheckBox,
  Dropdown: lightDropdown,
  AppModal: lightModal,
  ParallaxScrollView: lightParallaxScrollView,
};

export { Color };
export default lightTheme;
