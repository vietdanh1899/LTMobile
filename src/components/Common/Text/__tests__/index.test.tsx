import React from 'react';
import { fontFamily, fontWeight, fontSize } from '@themes/Font';
import {
  basicTest, testStyleProps, EnumComponent, getStyle, getComponent,
} from '@utils/testHelper';
import { lightTheme, Fonts } from '@themes';
import { Platform } from 'react-native';
import { Color } from '@themes/Theme';
import Text from '..';

const target = <Text />;
const type = EnumComponent.HOC;

describe('Text Component', () => {
  basicTest(target, type);

  it('should apply all style-related props', () => {
    const props = {
      margin: 10,
      marginVertical: 10,
      marginHorizontal: 10,
      marginTop: 10,
      marginBottom: 10,
      marginLeft: 10,
      marginRight: 10,
    };
    testStyleProps(target, props, type);
  });

  it('should apply all custom props', () => {
    /**
     * center
     */
    const testStyle = getStyle(target, { center: true }, type);
    expect(testStyle.textAlign).toBe('center');

    /**
     * fontFamily, fontWeight, fontSize
     */
    const props2 = {
      fontFamily: 'RobotoBold',
      fontWeight: 'thin',
      fontSize: 'large',
    };
    const testStyle2 = getStyle(target, props2, type);
    expect(testStyle2.fontFamily).toBe(fontFamily.RobotoBold);
    expect(testStyle2.fontWeight).toBe(fontWeight.thin);
    expect(testStyle2.fontSize).toBe(fontSize.large);

    /**
     * primary, secondary, success,  warning, error, color,
     */
    Platform.OS = 'android';
    const testStyle3 = getStyle(target, { primary: true }, type);
    expect(testStyle3.color).toBe(lightTheme.colors.platform.android.primary);
    const testStyle4 = getStyle(target, { secondary: true }, type);
    expect(testStyle4.color).toBe(lightTheme.colors.platform.android.secondary);
    const testStyle5 = getStyle(target, { success: true }, type);
    expect(testStyle5.color).toBe(lightTheme.colors.platform.android.success);
    const testStyle6 = getStyle(target, { error: true }, type);
    expect(testStyle6.color).toBe(lightTheme.colors.platform.android.error);
    const testStyle7 = getStyle(target, { warning: true }, type);
    expect(testStyle7.color).toBe(lightTheme.colors.platform.android.warning);
    const testStyle8 = getStyle(target, { color: Color.orange }, type);
    expect(testStyle8.color).toBe(Color.orange);
    jest.resetModules();

    Platform.OS = 'ios';
    const testStyle9 = getStyle(target, { primary: true }, type);
    expect(testStyle9.color).toBe(lightTheme.colors.platform.ios.primary);
    const testStyle10 = getStyle(target, { secondary: true }, type);
    expect(testStyle10.color).toBe(lightTheme.colors.platform.ios.secondary);
    const testStyle11 = getStyle(target, { success: true }, type);
    expect(testStyle11.color).toBe(lightTheme.colors.platform.ios.success);
    const testStyle12 = getStyle(target, { error: true }, type);
    expect(testStyle12.color).toBe(lightTheme.colors.platform.ios.error);
    const testStyle13 = getStyle(target, { warning: true }, type);
    expect(testStyle13.color).toBe(lightTheme.colors.platform.ios.warning);
    const testStyle14 = getStyle(target, { color: Color.orange }, type);
    expect(testStyle14.color).toBe(Color.orange);
    jest.resetModules();

    /**
     * underline, bold, italic
     */
    const testStyle15 = getStyle(target, { underline: true }, type);
    expect(testStyle15.textDecorationLine).toBe('underline');
    const testStyle16 = getStyle(target, { bold: true }, type);
    expect(testStyle16.fontWeight).toBe(Fonts.fontWeight.bold);
    const testStyle17 = getStyle(target, { italic: true }, type);
    expect(testStyle17.fontFamily).toBe(Fonts.fontFamily.RobotoItalic);

    /**
     * icon
     */
    // const component = mount(<Text icon={{ name: 'account' }} />);
    // const demo = component.findWhere((node) => node.prop('testID') === 'EIconText');
    // expect(demo.length).toBeGreaterThanOrEqual(1);

    const props3 = getComponent(target, { icon: { name: 'account' } }, type).props();
    expect(props3.testID).toBe('EIconText');
    expect(props3.row).toBe(true);
    expect(props3.rowReverse).not.toBe(true);

    /**
     * iconRight
     */
    const props4 = getComponent(target, { icon: { name: 'account' }, iconRight: true }, type).props();
    expect(props3.testID).toBe('EIconText');
    expect(props4.row).not.toBe(true);
    expect(props4.rowReverse).toBe(true);

    /**
     * iconContainerStyle
     */
    const props5 = getComponent(target, { icon: { name: 'account' }, iconContainerStyle: { color: Color.orange } }, type).childAt(0).props();
    expect(props3.testID).toBe('EIconText');
    expect(props5.style).toMatchObject({ color: Color.orange });

    /**
     * type
     */
    const testStyle18 = getStyle(target, { type: 'xTitle' }, type);
    expect(testStyle18).toMatchObject(lightTheme.Text.xTitle);
  });
});
