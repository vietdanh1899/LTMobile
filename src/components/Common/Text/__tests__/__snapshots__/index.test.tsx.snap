// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text Component should render without issues 1`] = `
<Themed.Text
  style={
    Object {
      "color": "#5760EB",
      "fontSize": 16,
    }
  }
/>
`;
